File: elixir/yacht/lib/yacht.ex
defmodule Yacht do
  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer
  def score(category, dice) do
    Enum.sort(dice) |> do_score(category)
  end

  defp do_score(dice, category) when category in ~w[ones twos threes fours fives sixes]a do
    die = from_category_to_integer(category)
    die * Enum.count(dice, &(&1 == die))
  end

  defp do_score([a, a, a, b, b] = dice, :full_house) when a != b, do: Enum.sum(dice)
  defp do_score([a, a, b, b, b] = dice, :full_house) when a != b, do: Enum.sum(dice)
  defp do_score([a, a, a, a, _], :four_of_a_kind), do: 4 * a
  defp do_score([_, a, a, a, a], :four_of_a_kind), do: 4 * a
  defp do_score([1, 2, 3, 4, 5], :little_straight), do: 30
  defp do_score([2, 3, 4, 5, 6], :big_straight), do: 30
  defp do_score([a, a, a, a, a], :yacht), do: 50
  defp do_score([_, _, _, _, _] = dice, :choice), do: Enum.sum(dice)
  defp do_score([_, _, _, _, _], _category), do: 0

  defp from_category_to_integer(:ones), do: 1
  defp from_category_to_integer(:twos), do: 2
  defp from_category_to_integer(:threes), do: 3
  defp from_category_to_integer(:fours), do: 4
  defp from_category_to_integer(:fives), do: 5
  defp from_category_to_integer(:sixes), do: 6
end

File: zig/yacht/yacht.zig
pub const Category = enum {
    ones,
    twos,
    threes,
    fours,
    fives,
    sixes,
    full_house,
    four_of_a_kind,
    little_straight,
    big_straight,
    choice,
    yacht,
};

pub fn score(dice: [5]u3, category: Category) u32 {
    var counts: [7]u32 = .{ 0, 0, 0, 0, 0, 0, 0 };
    var sum: u32 = 0;
    for (dice) |die| {
        counts[die] += 1;
        sum += die;
    }

    var meta: [6]u32 = .{ 0, 0, 0, 0, 0, 0 };
    for (counts) |count| {
        meta[count] += 1;
    }

    return switch (category) {
        .ones => counts[1],
        .twos => counts[2] * 2,
        .threes => counts[3] * 3,
        .fours => counts[4] * 4,
        .fives => counts[5] * 5,
        .sixes => counts[6] * 6,
        .full_house => if (meta[2] > 0 and meta[3] > 0) sum else 0,
        .four_of_a_kind => for (counts, 0..) |count, die| {
            if (count > 3) break @as(u32, @intCast(die)) * 4;
        } else 0,
        .little_straight => for (counts, 0..) |count, die| {
            if (die > 0 and die < 6 and count < 1) break 0;
        } else 30,
        .big_straight => for (counts, 0..) |count, die| {
            if (die > 1 and count < 1) break 0;
        } else 30,
        .choice => sum,
        .yacht => if (meta[5] > 0) 50 else 0,
    };
}
